staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -147,8 +147,13 @@ func (pq waitForPriorityQueue) Peek() interface{} {
 }
 
 // ShutDown stops the queue. After the queue drains, the returned shutdown bool
-// on Get() will be true. This method may be invoked more than once.
+// on Get() will be true. The assumption here is that every item put on the
+// queue will also finish processing at some point in time, hence: no need to
+// use a timeout. This method may be invoked more than once.
 func (q *delayingType) ShutDown() {
+	for q.Processing() {
```

Comment:
This doesn't seem like the right way to do this, what is to prevent people from adding more items?

I don't know how the delaying queue works (I wrote the non-delaying one) but generally a shut down should make sure that callers attempting to add new items to the queue get an error or at least don't actually add something to the queue.

Reply:
Indeed. I've updated my PR to have the verification of the `processing` length happen after we set `shuttingDown`, as such the queue will stop new items from being accepted but `ShutDown` will still wait until all items have been processed before returning - and thus have the controller process wait for the `syncHandler` to finish for all items before terminating.